llvm/lib/Target/DirectX/DXILPrettyPrinter.cpp

```diff
@@ -10,23 +10,235 @@
+      // TODO: dxc would print "ras" here. Can/should this happen?
+      llvm_unreachable("RTAccelerationStructure printing is not implemented");
+    }
+  }
+};
+
+struct FormatBindingID
+    : public llvm::FormatAdapter<const dxil::ResourceInfo &> {
+  explicit FormatBindingID(const dxil::ResourceInfo &RI)
+      : llvm::FormatAdapter<const dxil::ResourceInfo &>(RI) {}
+
+  void format(llvm::raw_ostream &OS, StringRef Style) override {
+    OS << getRCPrefix(Item.getResourceClass()).upper()
+       << Item.getBinding().RecordID;
+  }
+};
+
+struct FormatBindingLocation
+    : public llvm::FormatAdapter<const dxil::ResourceInfo &> {
+  explicit FormatBindingLocation(const dxil::ResourceInfo &RI)
+      : llvm::FormatAdapter<const dxil::ResourceInfo &>(RI) {}
+
+  void format(llvm::raw_ostream &OS, StringRef Style) override {
+    const auto &Binding = Item.getBinding();
+    OS << getRCPrefix(Item.getResourceClass()) << Binding.LowerBound;
+    if (Binding.Space)
+      OS << ",space" << Binding.Space;
+  }
+};
+
+struct FormatBindingSize
+    : public llvm::FormatAdapter<const dxil::ResourceInfo &> {
+  explicit FormatBindingSize(const dxil::ResourceInfo &RI)
+      : llvm::FormatAdapter<const dxil::ResourceInfo &>(RI) {}
+
+  void format(llvm::raw_ostream &OS, StringRef Style) override {
+    uint32_t Size = Item.getBinding().Size;
+    if (Size == std::numeric_limits<uint32_t>::max())
+      OS << "unbounded";
+    else
+      OS << Size;
+  }
+};
+
+} // namespace
+
+static void prettyPrintResources(raw_ostream &OS, const DXILResourceMap &DRM,
                                  const dxil::Resources &MDResources) {
   // Column widths are arbitrary but match the widths DXC uses.
   OS << ";\n; Resource Bindings:\n;\n";
-  OS << formatv("; {0,-30} {1,10} {2,7} {3,11} {4,7} {5,14} {6,16}\n",
+  OS << formatv("; {0,-30} {1,10} {2,7} {3,11} {4,7} {5,14} {6,9}\n",
                 "Name", "Type", "Format", "Dim", "ID", "HLSL Bind", "Count");
   OS << formatv(
-      "; {0,-+30} {1,-+10} {2,-+7} {3,-+11} {4,-+7} {5,-+14} {6,-+16}\n", "",
+      "; {0,-+30} {1,-+10} {2,-+7} {3,-+11} {4,-+7} {5,-+14} {6,-+9}\n", "",
       "", "", "", "", "", "");
 
+  // TODO: Do we want to sort these by binding or something like that?
+  for (auto [_, RI] : DRM) {
+    dxil::ResourceClass RC = RI.getResourceClass();
+    assert((RC != dxil::ResourceClass::CBuffer || !MDResources.hasCBuffers()) &&
+           "Old and new cbuffer representations can't coexist");
+    assert((RC != dxil::ResourceClass::UAV || !MDResources.hasUAVs()) &&
+           "Old and new UAV representations can't coexist");
+
+    StringRef Name(RI.getName());
+    StringRef Type(getRCName(RC));
+    StringRef Format(getFormatName(RI));
+    FormatResourceDimension Dim(RI);
+    FormatBindingID ID(RI);
+    FormatBindingLocation Bind(RI);
+    FormatBindingSize Count(RI);
+    OS << formatv("; {0,-30} {1,10} {2,7} {3,11} {4,7} {5,14} {6,9}\n",
```

Comment:
Neat use of the ` llvm::FormatAdapter`. Bit of a overkill for the simple cases but it looks elegant. ;).